File: dotnet/src/Microsoft.AutoGen/Core/TypePrefixSubscriptionAttribute.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// TypePrefixSubscriptionAttribute.cs

using Microsoft.AutoGen.Contracts;

namespace Microsoft.AutoGen.Core;

[AttributeUsage(AttributeTargets.All)]
public class TopicPrefixSubscriptionAttribute(string topic) : Attribute, IUnboundSubscriptionDefinition
{
    public string Topic { get; } = topic;

    public ISubscriptionDefinition Bind(AgentType agentType)
    {
        return new TypePrefixSubscription(Topic, agentType);
    }
}

File: typescript/core/TypePrefixSubscriptionAttribute.ts
import { ISubscriptionDefinition } from "../contracts/ISubscriptionDefinition";
import { IUnboundSubscriptionDefinition } from "../contracts/IUnboundSubscriptionDefinition";
import { AgentType } from "../contracts/AgentType";
import { TypePrefixSubscription } from "./TypePrefixSubscription";

export class TypePrefixSubscriptionAttribute implements IUnboundSubscriptionDefinition {
  constructor(private readonly topic: string) {}

  bind(agentType: AgentType): ISubscriptionDefinition {
    return new TypePrefixSubscription(this.topic, agentType);
  }
}
